### PR TITLE
Modernize library building and linting workflows

### PR DESCRIPTION
**Pull request by aris-aiven**
Modernize the project's build and linting story with multiple enhancements:

- Transition to `pyproject.toml`: this allows defining project metadata, dependencies & linting rules in a single place.
   It allows conforms to python packaging best practices by removing `setup.py`.
- Add pre-commit configuration: this allows simple linting and formatting with a single commands in a unified way for 
  multiple platforms
- Update `mypy`/`ruff` dependencies: the library benefits from linting and type checking updates from ruff and mypy
- Pin `mypy` and `pre-commit` target version to python 3.8 as well to ensure backward compatibility

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -52,3 +52,7 @@ venv:
 	pip install -U pip && \
 	pip install -r requirements.txt && \
 	pip install -r requirements.dev.txt
+
+.PHONY: venv
```
Comment (exg) [inline Makefile:55]: s/venv/clean/

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -17,8 +17,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        # only use one version for the lint step
-        python-version: ["3.11"]
+        python-version: ["3.11", "3.12"]
```
Comment (exg) [inline .github/workflows/build.yml:20]: Adding coverage for python 3.12 seems unrelated to transitioning to `pyproject.toml`.

Comment (aris-aiven) [reply .github/workflows/build.yml]: Indeed - reverted the change

File: pyproject.toml [hunk 3]
```diff
@@ -2,6 +2,84 @@
...
+    "Development Status :: 5 - Production/Stable",
+    "Intended Audience :: Developers",
+    "Intended Audience :: Information Technology",
+    "Intended Audience :: System Administrators",
+    "License :: OSI Approved :: Apache Software License",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
```
Comment (giacomo-alzetta-aiven) [inline pyproject.toml:24]: Same as above, I'd move python3.12 introduction stuff in its own PR

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -63,7 +59,7 @@ jobs:
     strategy:
       max-parallel: 4
       matrix:
-        python-version: ["3.11", "3.10", "3.9", "3.8"]
+        python-version: ["3.12", "3.11", "3.10", "3.9", "3.8"]
```
[the remark this answers is not in the document]
Comment (aris-aiven) [reply .github/workflows/build.yml]: Moved to https://github.com/Aiven-Open/rohmu/pull/178

**Comment by aris-aiven** [comment]
Moved testing with Python 3.12 to https://github.com/Aiven-Open/rohmu/pull/178

File: pyproject.toml [hunk 5]
```diff
@@ -2,6 +2,83 @@
...
+    "httplib2",
+    "oauth2client",
+    "paramiko",
+    "pydantic < 2",
+    "python-dateutil",
+    "python-swiftclient",
+    "requests",
+    # Kept in sync with Fedora versions
+    "python-snappy==0.6.1",
+    "zstandard==0.21.0",
```
Comment (kmichel-aiven) [inline pyproject.toml:42]: I don't think it should pin versions here in the non-dev dependencies.
(The pinned versions where in the requirements.dev.txt and not in the setup.cfg).

Comment (aris-aiven) [reply pyproject.toml:42]: Reverting the "silent" change to be strictly compatible with the previous format.

To be fair, I don't think the "dev" group is the right place to pin these dependencies, they would be better suited to a "compat" group. Out of scope for this PR nonetheless.

**Review by giacomo-alzetta-aiven — APPROVED** [review]
LGTM

I'm leaving some time to @kmichel-aiven  and @exg  before merging since they had requested some changes before.

**Comment by aris-aiven** [comment]
Pinning the python version for the Sphinx build job is necessary to avoid errors with the default version provided by different runners.  The setup-python actions docs (https://github.com/actions/setup-python) recommend always setting Python version explicitly.

**Review by exg — APPROVED** [review]

**Review by kmichel-aiven — APPROVED** [review]
